### DISCUSSION TITLE
S3 Cache Control Header

### DISCUSSION THREAD

**Discussion by sGerli** [directus/directus#7834] [Archive]
I think there should be a way to set the default S3 Cache Control metadata, maybe through an env variable. This would help reduce client calls to the S3 bucket or to Cloudfront when its used directly (we don't use Directus' media api because some of our files are very large and it could get expensive really fast).

I don't think this is currently implemented, but correct me if I'm wrong.

**Comment by rijkvanzanten** [comment]
All environment variables that are configured for the s3 storage adapter are passed along as a camelcased object to AWS SDK S3 (https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html). If the configuration flag you're looking for is supported in there, you should be able to configure it within Directus 👍🏻

**Reply by sGerli** [reply]
So far I've tried adding `STORAGE_AMAZON_PARAMS=json:{"CacheControl": "must-revalidate, max-age=3600"}` to my env and it doesn't seem to be working.

According to AWS S3 SDK that should work https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#:~:text=Calling%20the%20copyObject%20operation

Any ideas about what I could be doing wrong?

**Reply by azrikahar** [reply]
Directus does have [type casting for env variables](https://docs.directus.io/reference/environment-variables/#environment-syntax-prefix), but `json:` unfortunately isn't one of it.

For nesting (what you want to achieve), we should use [this instead](https://docs.directus.io/reference/environment-variables/#type-casting-and-nesting), specifically the double underscore. So `STORAGE_AMAZON_PARAMS__CACHE_CONTROL="must-revalidate, max-age=3600"` _should_ achieve what you need, but I have suspicion it might still not work because seems like the sdk wants `CacheControl` (Pascal case) instead of `cacheControl` (Camel case, which is what the current code does):

https://github.com/directus/directus/blob/5b59bce3b7497a8779d58eee269cd53e82fe1ee8/api/src/utils/get-config-from-env.ts#L41

Do let us know how it goes. If the suggestion still does not work, then it's likely a "aws sdk expects PascalCase" issue. 

Bad news is yet another irregularity with 3rd party sdk configs like #8569, good news is `camelcase` package [supports Pascal Case via a boolean flag](https://www.npmjs.com/package/camelcase), so technically it'll make things easier, but just need to think of a way to extend `get-config-from-env` that makes sense.

**Reply by rijkvanzanten** [reply]
@sGerli Isn't the cache control header controlled in the bucket configuration, instead of on a file-per-file basis? 🤔

**Reply by sGerli** [reply]
> @sGerli Isn't the cache control header controlled in the bucket configuration, instead of on a file-per-file basis? 🤔 

Sadly no, that was the first possibility I explored, but it seems to be on a per-file basis. Some people have found a way to do it using an lambda function between s3 an cloudfront but IMO that's no the right way to do it.

I'll try @azrikahar s suggestion but I agree with him that it's probably not going to work due to the camelCase issue

**Reply by rijkvanzanten** [reply]
@azrikahar JSON strings are actually tried as JSON by default for any value, so this is actually possible 🙂 

@sGerli Try `STORAGE_AMAZON_PARAMS={"CacheControl":"must-revalidate, max-age=3600"}` instead, I believe it should work like that (though haven't tested it myself) 👍🏻

**Reply by sGerli** [reply]
@rijkvanzanten that did work, thanks 🎉

**Reply by azrikahar** [reply]
@rijkvanzanten thanks for the correction!

@sGerli glad it worked 👍
